### DISCUSSION TITLE
Is it possible to set the Z Value on an individual spot item?

### DISCUSSION THREAD

**Discussion by pjm4github** [pyqtgraph/pyqtgraph#2433] [Q&A]
I have thousands of scatter plot items that I need to stack so that I can enable the user to select the items of interest with their mouse. 

The scatter plot represents a graph with connected nodes. When the parent node is selected I want to pull all the children nodes forward in the scatter plot. 

When I try and set the Z value on an individual SpotItem it sets the Z value for all spots in the ScatterPlotItem.
How do I reach down into the canvas item and for the spot item and get/ set its individual Z value? 

Is that possible?

In my clicked callback of the GraphItem subclass I thought I could set it there. This obviously sets the zValue for the entire plot. 

Any ideas? 

```

class GraphPallet(GraphItem):

....

    def clicked(self, scatter, pts):
        print("clicked")
        data_list = scatter.data.tolist()
        mypoint = [tup for tup in data_list if pts[0] in tup][0]
        mypoint_index = data_list.index(mypoint)
        for item_dict in scatter.data:
            item_dict['item']._plot.setZValue(0.0)
        # Set this item to the front (on top) of the other items
        scatter.data[mypoint_index]['item']._plot.setZValue(10.0)

```

**Comment by ixjlyons** [comment]
Because of the way `ScatterPlotItem` works, it's not really possible to change the stacking order of points without going to somewhat extreme measures (only way I can think of is changing the order of the input data and re-generating).

Maybe you could add an additional "selected" item to the scene that you move and show/hide in response to selection/deselection events.

**Reply by pjm4github** [reply]
OK I see. It looks like the items are rendered as pixmaps at the base level.  Im not sure what the SymbolAtlas class does - but I'm curious why the QGraphicsScene was completely abandoned in favor of this method. (I'm probably completely lost)

Anyway - I think what I will do, once the user selects a node (spot) to examine, is to then remove the nearest neighbors in the selected tree then add them back in once the user selects a node (spot) to examine.  I've read that the "stacking order" is set by the order in which the spots are added so that should be what I need.  Is that going to burn me in some other way? 

Thanks!

**Reply by ixjlyons** [reply]
My understanding is it's a performance optimization - I'd imagine each spot being a graphics item might get sluggish for lots of points. Generating and caching a picture makes the `paint()` calls cheap and moving around in the plot view smooth.

And I think what you describe should work - changing the order of points in the arrays passed to `ScatterPlotItem` will change their stacking order - but it probably won't be very snappy for ~thousands of points. For example, if you wanted to raise the spot under the cursor on hover with this method, the interface would probably feel pretty slow. That's why I was suggesting adding another item to the scene to represent a selected spot, but I'm not sure what exactly you're looking to do.

**Comment by j9ac9k** [comment]
@pjm4github if I was going to do this, I would do multiple overlapping scatter plots, one of the scatter plots would have the symbol(s) I wanted on top of the others, and that scatter plot would have a higher Z value than the other one.

The scatter plot stuff is pretty complex, since we use QPainter.drawPixmapFragments we feed in all the pixmaps at once, I'm not sure how internally order is determined but I think trying to tinker with modifying how scatter plot does order placement.

**Reply by pjm4github** [reply]
Interesting idea. My scatter plot data is pretty much a high tree.  Its based on networkx digraph which is what attracted me to pyqtgraph to begin with. The tree is like the figure attached. When its laid out the leaf nodes are too dense for picking with a mouse. 
I'm not sure how to break this apart without having 100s of scatter plots. 
My goal here is to let the user select one of the blue nodes (for example) then bring the parent and children nodes forward to allow easy picking. 
In my case the graph I use has several thousand nodes and 10s of parents.  

I'll just live with the picker problem for now. 

I'm very impressed with the performance of the UI with pyqtgraph.

![image](https://user-images.githubusercontent.com/3688079/192879086-8617aa63-2563-46b6-b502-4a70b931642c.png)

Thanks again for your comments!

**Reply by j9ac9k** [reply]
What's wrong with just zooming in on the area of the denser nodes? as long as pxMode is True (default behavior), eventually you will see daylight between the nodes

**Reply by pjm4github** [reply]
Yes thats what I do now. Unfortunately the navigation is clunky because the text items with the nodes doesn't scale and you loose the big picture. In order to zoom in close enough I can't even differentiate which node is connected to my parent. It's a range resolution issue.

**Reply by j9ac9k** [reply]
Oh that's interesting the text doesn't scale, that probably _should_ happen.

**Reply by j9ac9k** [reply]
If you can open an issue, with a code to reproduce the issue of text not rescaling with the scatter plot in pxMode, that would be very helpful!

**Reply by pjm4github** [reply]
Will do.
